http-client-jdk/src/main/java/io/micronaut/http/client/jdk/AbstractJdkHttpClient.java

```diff
@@ -0,0 +1,277 @@
+
+    protected final LoadBalancer loadBalancer;
+    protected final HttpVersionSelection httpVersion;
+    protected final HttpClientConfiguration configuration;
+    protected final String contextPath;
+    protected final HttpClient client;
+    protected final CookieManager cookieManager;
+    protected final RequestBinderRegistry requestBinderRegistry;
+    protected final String clientId;
+    protected final ConversionService conversionService;
+    protected final JdkClientSslBuilder sslBuilder;
+    private final Logger log;
+    protected MediaTypeCodecRegistry mediaTypeCodecRegistry;
+
+    /**
+     * @param log                    the logger to use
+     * @param loadBalancer           The {@link LoadBalancer} to use for selecting servers
+     * @param httpVersion            The {@link HttpVersionSelection} to prefer
+     * @param configuration          The {@link HttpClientConfiguration} to use
+     * @param contextPath            The base URI to prepend to request uris
+     * @param mediaTypeCodecRegistry The {@link MediaTypeCodecRegistry} to use for encoding and decoding objects
+     * @param requestBinderRegistry  The request binder registry
+     * @param clientId               The client id
+     * @param conversionService      The {@link ConversionService}
+     * @param sslBuilder             The {@link JdkClientSslBuilder} for creating an {@link javax.net.ssl.SSLContext}
+     */
+    protected AbstractJdkHttpClient(
+        Logger log,
+        LoadBalancer loadBalancer,
+        HttpVersionSelection httpVersion,
+        HttpClientConfiguration configuration,
+        String contextPath,
+        MediaTypeCodecRegistry mediaTypeCodecRegistry,
+        RequestBinderRegistry requestBinderRegistry,
+        String clientId,
+        ConversionService conversionService,
+        JdkClientSslBuilder sslBuilder
+    ) {
+        this.log = configuration.getLoggerName().map(LoggerFactory::getLogger).orElse(log);
+        this.loadBalancer = loadBalancer;
+        this.httpVersion = httpVersion;
+        this.configuration = configuration;
+        this.mediaTypeCodecRegistry = mediaTypeCodecRegistry;
+        this.requestBinderRegistry = requestBinderRegistry;
+        this.clientId = clientId;
+        this.conversionService = conversionService;
+        this.cookieManager = new CookieManager();
+        this.sslBuilder = sslBuilder;
+
+        if (StringUtils.isNotEmpty(contextPath)) {
+            if (contextPath.charAt(0) != '/') {
+                contextPath = '/' + contextPath;
+            }
+            this.contextPath = contextPath;
+        } else {
+            this.contextPath = null;
+        }
+
+        HttpClient.Builder builder = HttpClient.newBuilder();
+        configuration.getConnectTimeout().ifPresent(builder::connectTimeout);
+
+        builder
+            .version(httpVersion != null && httpVersion.isAlpn() ? HttpClient.Version.HTTP_2 : HttpClient.Version.HTTP_1_1)
```

Comment:
there should be more checks here

- h2c is not supported by the jdk (check plaintextMode)
- the user could enable alpn without h2, in which case http1 should be used

the first may be a bit annoying. if you set http2 in the config using the old http version field, it will enable h2c, so this would error. but i think it's correct to error here.

Reply:
Based on @yawkat probably best to move it to a protected method someone may override if we miss anything.

Reply:
i dont think it's necessary, it is already possible to set the right config to get the jdk client version you want. i just think the heuristic should be a bit different to closer match the netty client

Reply:
Like this you mean? https://github.com/micronaut-projects/micronaut-core/pull/8441/commits/9174a0026a2f5b2ecb375f4ced4da30accf496e3

I'm a bit lost in the many, many, many ways we have for configuring this...

Reply:
Like this? https://github.com/micronaut-projects/micronaut-core/pull/8441/commits/756045fa30b246615f8ed23b06532dc742745a02